wire/netaddressv2.go

```diff
@@ -0,0 +1,588 @@
+		addr := &torv2Addr{}
+		addr.netID = torv2
+		if decodedSize != uint64(torv2Size) {
+			return ErrInvalidAddressSize
+		}
+
+		if err := readElement(r, &addr.addr); err != nil {
+			return err
+		}
+
+		na.Port, err = binarySerializer.Uint16(r, bigEndian)
+		if err != nil {
+			return err
+		}
+
+		na.Addr = addr
+	case torv3:
+		addr := &torv3Addr{}
+		addr.netID = torv3
+		if decodedSize != uint64(TorV3Size) {
+			return ErrInvalidAddressSize
+		}
+
+		if err := readElement(r, &addr.addr); err != nil {
+			return err
+		}
+
+		na.Port, err = binarySerializer.Uint16(r, bigEndian)
+		if err != nil {
+			return err
+		}
+
+		// BIP-155 does not specify to validate the public key here.
+		// bitcoind does not validate the ed25519 pubkey.
+		na.Addr = addr
+	case i2p:
+		addr := &i2pAddr{}
+		addr.netID = i2p
+		if decodedSize != uint64(i2pSize) {
+			return ErrInvalidAddressSize
+		}
+
+		if err := readElement(r, &addr.addr); err != nil {
+			return err
+		}
+
+		na.Port, err = binarySerializer.Uint16(r, bigEndian)
+		if err != nil {
+			return err
+		}
+
+		return ErrSkippedNetworkID
+	case cjdns:
+		addr := &cjdnsAddr{}
+		addr.netID = cjdns
+		if decodedSize != uint64(cjdnsSize) {
+			return ErrInvalidAddressSize
+		}
+
+		if err := readElement(r, &addr.addr); err != nil {
+			return err
+		}
+
+		na.Port, err = binarySerializer.Uint16(r, bigEndian)
+		if err != nil {
+			return err
+		}
+
+		return ErrSkippedNetworkID
+	}
+
+	return nil
+}
+
+// networkID represents the network that a given address is in. CJDNS and I2P
+// addresses are not included.
+type networkID uint8
+
+const (
+	// ipv4 means the following address is ipv4.
+	ipv4 networkID = iota + 1
+
+	// ipv6 means the following address is ipv6.
+	ipv6
+
+	// torv2 means the following address is a torv2 hidden service address.
+	torv2
+
+	// torv3 means the following address is a torv3 hidden service address.
+	torv3
+
+	// i2p means the following address is an i2p address.
+	i2p
+
+	// cjdns means the following address is a cjdns address.
+	cjdns
+)
+
+const (
+	// ipv4Size is the size of an ipv4 address.
+	ipv4Size = 4
+
+	// ipv6Size is the size of an ipv6 address.
+	ipv6Size = 16
+
+	// torv2Size is the size of a torv2 address.
+	torv2Size = 10
+
+	// TorV3Size is the size of a torv3 address in bytes.
+	TorV3Size = 32
+
+	// i2pSize is the size of an i2p address.
+	i2pSize = 32
+
+	// cjdnsSize is the size of a cjdns address.
+	cjdnsSize = 16
+)
+
+const (
+	// TorV2EncodedSize is the size of a torv2 address encoded in base32
+	// with the ".onion" suffix.
+	TorV2EncodedSize = 22
+
+	// TorV3EncodedSize is the size of a torv3 address encoded in base32
+	// with the ".onion" suffix.
+	TorV3EncodedSize = 62
+)
+
+// isKnownNetworkID returns true if the networkID is one listed above and false
+// otherwise.
+func isKnownNetworkID(netID uint8) bool {
+	return uint8(ipv4) <= netID && netID <= uint8(cjdns)
+}
+
+type ipv4Addr struct {
+	addr  [ipv4Size]byte
+	netID networkID
```

Comment:
`Network()` returns a string, and is only provided to make the structs conform to the net.Addr interface, so we'd have to convert back and forth to an int which I wanted to avoid. I think a `newIPV4Addr` and related constructors might work, but there may have been a reason I'm forgetting that I didn't do this. How would addr size fit in here?